8302819: Remove JAR Index

eirbjo opened the pull request:
This PR removes the JAR index feature from the runtime:

- `URLClassPath` is  updated to remove the `enableJarIndex` system property and any code which would be called when this property was `true`
- The `JarIndex` implementation class is moved into `jdk.jartool` module.
- The `InvalidJarIndexError` exception class is removed because it falls out of use
- The test `test/jdk/sun/misc/JarIndex/metaInfFileNames/Basic.java` is removed because it depends on the JarIndex feature being present
- The test `test/jdk/sun/misc/JarIndex/JarIndexMergeForClassLoaderTest.java` is removed because it depends on the JarIndex feature being present
- The test `test/jdk/sun/misc/JarIndex/JarIndexMergeTest.java` is removed because it end up being the only caller of the  JarIndex.merge feature
- All `JarIndex` methods/constructors which are not used by the `jar -i` implementation are removed. 
- `JarIndex` is given package-private access.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change requires CSR request [JDK-8305597](https://bugs.openjdk.org/browse/JDK-8305597) to be approved
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issues
 * [JDK-8302819](https://bugs.openjdk.org/browse/JDK-8302819): Remove JAR Index
 * [JDK-8305597](https://bugs.openjdk.org/browse/JDK-8305597): Remove JAR Index (**CSR**)


### Reviewers
 * [Mandy Chung](https://openjdk.org/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to [79b19e94](https://git.openjdk.org/jdk/pull/13158/files/79b19e94def64fc3791a10d9d29890e35f1e2e8b)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [79b19e94](https://git.openjdk.org/jdk/pull/13158/files/79b19e94def64fc3791a10d9d29890e35f1e2e8b)
 * [Lance Andersen](https://openjdk.org/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**) ⚠️ Review applies to [79b19e94](https://git.openjdk.org/jdk/pull/13158/files/79b19e94def64fc3791a10d9d29890e35f1e2e8b)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13158/head:pull/13158` \
`$ git checkout pull/13158`

Update a local copy of the PR: \
`$ git checkout pull/13158` \
`$ git pull https://git.openjdk.org/jdk.git pull/13158/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13158`

View PR using the GUI difftool: \
`$ git pr show -t 13158`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13158.diff">https://git.openjdk.org/jdk/pull/13158.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13158#issuecomment-1487368010)

jaikiran commented:
Hello Eirik,

> Note that this PR does not aim to remove the ability for the jar tool to produce JAR files with indexes. Such a removal could be handled separately from this PR.

My understanding of https://bugs.openjdk.org/browse/JDK-8302819 is that we are removing the JAR Index since it is no longer usable in recent versions.

The work, that by default disabled support for jar indexes in URLClassLoader https://bugs.openjdk.org/browse/JDK-8273473 did not deprecate the `jar -i` option and explicitly noted that it would continue to work the way it used to.
Perhaps now is the time to add a deprecation/warning message when `jar -i` option is used? JDK-8302819 does note this as a possibility as part of this change:
> The jar -i (--generate-index=FILE) option can continue to generate the JAR index or the option can be changed to just emit a warning.

jaikiran commented:
> 
>  The `InvalidJarIndexError` exception class is removed because it falls out of use

Since `Throwable` classes are serializable, I looked up the serialization spec to see if this removal would have any compatibility issues that need to be accounted for. The specification https://docs.oracle.com/en/java/javase/20/docs/specs/serialization/version.html#compatible-changes states that removing classes is a compatible change, so this should be fine. Plus, this exception appears to be thrown only when index file is being processed (as the name suggests).

jaikiran commented:
Sorry my previous message wasn't clear.

> Hooking in a simple deprecation warning in this PR would propably not require a lot of extra review cycles, perhaps the opposite. So I'm happy to do that if that is what you meant

Yes, that's what I meant - I think we should add a warning message when that option is used, as part of these changes. I don't think we can completely remove that option in this release, since it wasn't deprecated (for removal) before, so it would be a good idea to start warning. Like you, I too would let Alan provide guidance on whether this should be done as part of this PR.

eirbjo commented:
> Sorry my previous message wasn't clear.

Thanks for clarifying, Jaikiran. I think our understanding is in sync now, which is always good!

eirbjo commented:
If we include the deprecation of `jar -i` in this PR, then the CSR for this PR would need to also cover that deprecation. Same is true for release notes. Not seeing any problem with that myself, but thought it would be good to take note here.

jaikiran commented:
> @jaikiran Do you agree with the move of `JarIndex` from `java.base` to `jdk.jartool`? Since it ends up only being used there, I thought it would be cleaner if we moved it. But the move is not strictly necessary. What's your thoughts?

With the changes in this PR, the `JarIndex` class would only be relevant/useful in the `jar` tool. So I think it makes sense to move it to the `jdk.jartool` module. Additionally you could even make it package-private there, instead of public. It would anyway need some updates to its javadoc and maybe remove any no longer used methods.

On a related note, I see that the `module-info.java` of `java.base` module has been updated to export `sun.security.action` package to the `jdk.jartool`. I think that is because the (now moved) `JarIndex` uses the `sun.security.action.GetPropertyAction.privilegedGetProperty` utility method to read a system property. I don't think we should be exporting that package. Instead, the `JarIndex` class can be updated to do `AccessController.doPrivileged(...)`.

eirbjo commented:
> Do you know if the `jar` tool allows running with a `SecurityManager`? If not, we could perhaps simply use `System.getProperty`?

I ended up feeling opimistic and replaced the current code with just `System.getProperty`. If that is any concern doing that, let me know and I can use `AccessController` instead.

eirbjo commented:
Regarding `JarIndex`:

>Additionally you could even make it package-private there, instead of public. It would anyway need some updates to its javadoc and maybe remove any no longer used methods.

This did not work because of the `JarIndexMergeTest` test which directly depends on `JarIndex`. Changing package of the test did not work because of split packages.

eirbjo commented:
@jaikiran made the good observation that the `JarIndex` class has various cruft left which is not needed for the `jar -i` implementation.

Since we at the moment do not know when `jar -i` will be removed (if ever), we would risk leaving a lot of unused code around if we chose to postpone this cleanup.

Becuse of this, I suggest we extend the scope of this PR to clean up `IndexJar` for use by `jdk.jartool` only.

jaikiran commented:
> > I don't think we should be exporting that package. Instead, the `JarIndex` class can be updated to do `AccessController.doPrivileged(...)`.
> 
> Nice, I'll try that!
> 
> Do you know if the `jar` tool allows running with a `SecurityManager`? If not, we could perhaps simply use `System.getProperty`?

I had a look at the existing code in the jar tool. It appears that it doesn't run with a SecurityManager - there's direct calls to `System.getProperty()` in some places in the code. So I think just calling `System.getProperty()` should be fine. When this PR goes into a published state for review, I'm sure others who have more knowledge about this area will help decide.

eirbjo commented:
@jaikiran 

Looking into how the deprecation warning could be implemented, it seems `jar` has some existing infrastructure for deprecation warnings, so the deprecation could be implemented with the following small patch:

```
Subject: [PATCH] Remove JarIndex default constructor
---
Index: src/jdk.jartool/share/classes/sun/tools/jar/Main.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/jdk.jartool/share/classes/sun/tools/jar/Main.java b/src/jdk.jartool/share/classes/sun/tools/jar/Main.java
--- a/src/jdk.jartool/share/classes/sun/tools/jar/Main.java	(revision fee0da27be832a7413886ddd2cfd5586bfd2e37a)
+++ b/src/jdk.jartool/share/classes/sun/tools/jar/Main.java	(date 1679654673667)
@@ -396,6 +396,9 @@
                     }
                 }
             } else if (iflag) {
+                if (!suppressDeprecateMsg) {
+                    warn(formatMsg("warn.flag.is.deprecated", "-i"));
+                }
                 String[] files = filesMap.get(BASE_VERSION);  // base entries only, can be null
                 genIndex(rootjar, files);
             } else if (dflag) {
```

This gives the following warning message on use:

```
% build/macosx-x86_64-server-release/images/jdk/bin/jar --generate-index file.jar
Warning: The -i option is deprecated, and is planned for removal in a future JDK release
```

Which may be overridden by using `-XDsuppress-tool-removal-message`:

```
% build/macosx-x86_64-server-release/images/jdk/bin/jar --generate-index file.jar -XDsuppress-tool-removal-message

```

jaikiran commented:
Hello Eirik, thank you for the updates to the PR.
I see that Alan suggested that we take up the `jar -i` deprecation in a separate PR. I think that simplifies the work in this PR. I'll run some tests with these changes and do another review soon, but you don't have to wait for it, if you want to move this PR out of draft status.

AlanBateman commented:
/label remove compiler

AlanBateman commented:
> I see that Alan suggested that we take up the `jar -i` deprecation in a separate PR. I think that simplifies the work in this PR.

Yes, there are options for `jar -i` that include "leave it as is", have the option create the index and emit a warning that the feature is deprecated, or remove the index creation and emit a warning that it is ignored. I don't think we should change "jar -i` to fail with an error as that would break existing scripts. None of these directions impact the JAR file spec or the API docs, it's just the jar tool so I think can be separated to another PR.

jaikiran commented:
Hello Eirik, I ran our tier tests against the current state of this PR and they came back fine.

With the removal of support for `META-INF/INDEX.LIST`, this PR rightly removes the related test cases. I think we should introduce a new test class with some test methods which verifies that the `java.util.jar.JarFile` and `java.util.jar.JarInputStream` APIs continue to work fine when they are presented with a jar which has a `META-INF/INDEX.LIST`. Some such tests would be to use these APIs to load:

 - A jar which has a META-INF/MANIFEST.MF and a META-INF/INDEX.LIST
 - A jar which has a META-INF/INDEX.LIST but no manifest file
 - A signed jar which has a META-INF/INDEX.LIST

Adding these tests I think will help us verify that these APIs  and also exercise the code in these classes where we have checks for `META-INF/INDEX.LIST` entry. Is that something you could consider adding as part of this PR?

eirbjo commented:
> Adding these tests I think will help us verify that these APIs and also exercise the code in these classes where we have checks for `META-INF/INDEX.LIST` entry. Is that something you could consider adding as part of this PR?

Yes, I agree this should be done. For now, I've added a note in the "Outstanding code work" section at the top of this PR.

AlanBateman commented:
> > Adding these tests I think will help us verify that these APIs and also exercise the code in these classes where we have checks for `META-INF/INDEX.LIST` entry. Is that something you could consider adding as part of this PR?
> 
> Yes, I agree this should be done. For now, I've added a note in the "Outstanding code work" section at the top of this PR.

Adding a test would be good. In the mean-time, I think it would be good to get the review started to flush out any concerns about removing this index.

LanceAndersen commented:
> > I see that Alan suggested that we take up the `jar -i` deprecation in a separate PR. I think that simplifies the work in this PR.
> 
> Yes, there are options for `jar -i` that include "leave it as is", have the option create the index and emit a warning that the feature is deprecated, or remove the index creation and emit a warning that it is ignored. I don't think we should change "jar -i` to fail with an error as that would break existing scripts. None of these directions impact the JAR file spec or the API docs, it's just the jar tool so I think can be separated to another PR.

Agree, we should handle this separately.  If we are thinking of emitting a warning, we probably should do that in = JDK 22  as that could break tests with golden files so we might want to give a heads up in JDK 21  before doing this

AlanBateman commented:
> If we are thinking of emitting a warning, we probably should do that in = JDK 22 as that could break tests with golden files so we might want to give a heads up in JDK 21 before doing this

The best option might be to just have `jar -i` work as before (as it does in the current proposal) but have it warn that -i may be changed to do nothing and/or be removed in a future release. If there are tests that looking for META/INDEX.list then they will work as before.

mlchung commented on src/java.base/share/classes/jdk/internal/loader/URLClassPath.java:
```diff
@@ -968,137 +924,15 @@ Resource getResource(final String name, boolean check) {
-            if (index != null) {
-                return null;
-            }
 
```
Nit: extra line can be removed.

eirbjo replied:
Thank to @mlchung,  line 936 is now history!

mlchung reviewed (commented):
Happy to see this work!  This change looks fine in general.    I agree that it's good to keep this issue to remove the runtime support and separate the `jar -i` change.   JDK tools don't run with security manager enabled and no need to wrap `System::getProperty` with `doPrivileged`.

AlanBateman commented:
/csr

AlanBateman commented on src/java.base/share/classes/java/util/jar/JarFile.java:
```diff
@@ -207,12 +207,17 @@ public class JarFile extends ZipFile {
 
+    /**
+     * The 'JAR index' feature has been removed, but we still need to
+     * process existing files which have this entry.
```
"but we still need ..."   Maybe better to say that that JarInputStream and the verification of signed JARs need to skip this attribute.

eirbjo replied:
I rephrased this to be more specific about which code still needs this. Agree "we" is not the best choice of subject here.

AlanBateman commented on src/jdk.jartool/share/classes/sun/tools/jar/Main.java:
```diff
@@ -69,14 +69,13 @@
 import jdk.internal.opt.CommandLine;
-import jdk.internal.util.jar.JarIndex;
+
 import java.time.LocalDateTime;
```
This can move up to the imports of the other java.* classes.

eirbjo replied:
Fixed.

AlanBateman reviewed (commented):
This looks good. I've added the "csr" label to the PR.

For "jar -i", I think we are converging on the option to just add a warning, meaning it will generated/update the index as before, but emit a warning that JAR Index is ignored since JDK 18. It doesn't matter if we do this in this PR or a separate PR. For CSR and release note purposes it would be a bit easier if it was this PR, would that be okay?

eirbjo commented:
> For CSR and release note purposes it would be a bit easier if it was this PR, would that be okay?

We seem to have wiggled a bit on this issue, but I'm ok with including it.

I have implemented the -i warning using the existing warnings strings already in place in `jar`. This means the following warning will be output:

```
% build/macosx-x86_64-server-release/images/jdk/bin/jar --generate-index file.jar
Warning: The -i option is deprecated, and is planned for removal in a future JDK release
```

Which may be overridden by using `-XDsuppress-tool-removal-message`:

```
% build/macosx-x86_64-server-release/images/jdk/bin/jar --generate-index file.jar -XDsuppress-tool-removal-message

```

I'm not sure we are planning to remove it though, so perhaps the current message is somewhat misleading? On the other hand, this message will allow us to remove it in the future, which is perhaps good?

AlanBateman commented:
> We seem to have wiggled a bit on this issue, but I'm ok with including it.

I think we've converged on `jar -i` work as before but print a warning. For the purposes of the CSR and the release note then I think it would be better to have the complete story in one issue rather than some of the story in one issue the two issues needing to cross reference each other. So thanks for going with that.

> 
> I have implemented the -i warning using the existing warnings strings already in place in `jar`. This means the following warning will be output:
> 
> ```
> % build/macosx-x86_64-server-release/images/jdk/bin/jar --generate-index file.jar
> Warning: The -i option is deprecated, and is planned for removal in a future JDK release
> ```

I think the warning should also communicate that the JAR index is also ignored, maybe something like this:

```
The JAR index (META-INF/INDEX.LIST) is ignored at run-time since JDK 18.
The --generate-index/-i option is deprecated - it may be ignored or removed in a future release.
```

mlchung commented:
> Is this repurposing of warn.flag.is.deprecated considered ok?

yes, that's not used.   It was a leftover from JDK-8204981 & JDK-8234596.

eirbjo commented:
I have drafted CSR https://bugs.openjdk.org/browse/JDK-8305597, but I'm not sure what to put in the _Specification_ section.

What is the specification change in this PR? The removal of the system property?  The adding of the warning?

I'm not seeing any API change as is probably more usual in a CSR.

AlanBateman commented:
> I have drafted CSR https://bugs.openjdk.org/browse/JDK-8305597, but I'm not sure what to put in the _Specification_ section.
> 
> What is the specification change in this PR? The removal of the system property? The adding of the warning?
> 
> I'm not seeing any API change as is probably more usual in a CSR.

I've edited it to add a Specification section with a summary of the changes to the JAR file spec. I also update the Solution section so that it provides a clearer summary of the changes.

LanceAndersen commented:
> > I have drafted CSR https://bugs.openjdk.org/browse/JDK-8305597, but I'm not sure what to put in the _Specification_ section.
> > What is the specification change in this PR? The removal of the system property? The adding of the warning?
> > I'm not seeing any API change as is probably more usual in a CSR.
> 
> I've edited it to add a Specification section with a summary of the changes to the JAR file spec. I also update the Solution section so that it provides a clearer summary of the changes.

I also updated the CSR with the revised JAR File Spec.   I will look over the changes and add myself as a reviewer

eirbjo commented:
> I've edited it to add a Specification section with a summary of the changes to the JAR file spec. I also update the Solution section so that it provides a clearer summary of the changes.

Thanks a lot Alan, as this CSR has deeper waters than what I've experienced before. Did you intend to remove the line break in the example warning output, collapsing it into one line?

mlchung reviewed (approved):
LGTM

eirbjo commented:
I added a draft release note https://bugs.openjdk.org/browse/JDK-8305603 as a sub-task of this PR's issue https://bugs.openjdk.org/browse/JDK-8303410

jaikiran reviewed (approved):
Hello Eirik, the changes look good to me. The new test that was planned for testing `JarFile` and `JarInputStream` when dealing with a jar containing a `META-INF/INDEX.LIST`, is that something that you want to be done as a separate PR/task? It's fine with me if you want that to be done separately.

AlanBateman commented:
> Thanks a lot Alan, as this CSR has deeper waters than what I've experienced before. Did you intend to remove the line break in the example warning output, collapsing it into one line?

I used the inline code formatting (3 backquotes) to make the output clear, but maybe it looks differently for you or maybe your comment was while I was editing? When you are ready, press "Finalize" and that will put in the CSR queue to review.

eirbjo commented:
> I used the inline code formatting (3 backquotes) to make the output clear, but maybe it looks differently for you or maybe your comment was while I was editing? 

It looks good now, so my comment was probably made during your edits.

> When you are ready, press "Finalize" and that will put in the CSR queue to review.

Done.

eirbjo commented:
> The new test that was planned for testing `JarFile` and `JarInputStream` when dealing with a jar containing a `META-INF/INDEX.LIST`, is that something that you want to be done as a separate PR/task? It's fine with me if you want that to be done separately.

@jaikiran 

I was looking into creating such a test yesterday. For `JarInputStream`, I was able to write a tests which expects a certain order of entries. This could made broken by disabling the `META-INF/INDEX.LIST` check in `JarInputStream.getNextEntry`.

With `JarVerifier`, I'm struggling to create a scenario which excercises the `META-INF/INDEX.LIST` check. It seems to me this code is simply unreachable, even after moving the `META-INF/INDEX.LIST` file first in the JAR.

If the code is properly unreachable, then maybe we should consider removing it?

In any case, since I feel the current test is somewhat dubious, I checked it into another branch here:

https://github.com/eirbjo/jdk/blob/remove-jar-index-test/test/jdk/java/util/jar/JarFile/JarIndexLegacyJars.java

This is perhaps better than having no tests. I'm not feeling sure about this, please let me know what you think. Perhaps @AlanBateman also has an opinion on this?

LanceAndersen commented:
> > The new test that was planned for testing `JarFile` and `JarInputStream` when dealing with a jar containing a `META-INF/INDEX.LIST`, is that something that you want to be done as a separate PR/task? It's fine with me if you want that to be done separately.
> 
> @jaikiran
> 
> I was looking into creating such a test yesterday. For `JarInputStream`, I was able to write a tests which expects a certain order of entries. This could made broken by disabling the `META-INF/INDEX.LIST` check in `JarInputStream.getNextEntry`.
> 
> With `JarVerifier`, I'm struggling to create a scenario which excercises the `META-INF/INDEX.LIST` check. It seems to me this code is simply unreachable, even after moving the `META-INF/INDEX.LIST` file first in the JAR.
> 
> If the code is properly unreachable, then maybe we should consider removing it?
> 
> In any case, since I feel the current test is somewhat dubious, I checked it into another branch here:
> 
> https://github.com/eirbjo/jdk/blob/remove-jar-index-test/test/jdk/java/util/jar/JarFile/JarIndexLegacyJars.java
> 
> This is perhaps better than having no tests. I'm not feeling sure about this, please let me know what you think. Perhaps @AlanBateman also has an opinion on this?

I am not convinced we need a test but to your point Jai, Eirik, lets handle this as a separate PR/Issue and work through what we are hoping to achieve/validate

eirbjo commented:
> I am not convinced we need a test but to your point Jai, Eirik, lets handle this as a separate PR/Issue and work through what we are hoping to achieve/validate

Good. Let's agree to handle testing of dusty 'index jars' separately from this PR.

AlanBateman commented:
One additional thing we could do is update the usage output for -i/--generate-index to say that the option is deprecated. Looking at `jdeps -help` output to see this for the deprecated -P/--profile option.

eirbjo commented:
> One additional thing we could do is update the usage output for -i/--generate-index to say that the option is deprecated. Looking at `jdeps -help` output to see this for the deprecated -P/--profile option.

Implemented as suggested. The --help output now looks like:

```
  -i, --generate-index=FILE  Generate index information for the specified jar
                             archives. This option is deprecated and may be 
                             removed in a future release.
```

AlanBateman commented:
I've updated the CSR to align it with the proposed solution. Please review again and if you happy, press "Finalize".

eirbjo commented:
> I've updated the CSR to align it with the proposed solution. Please review again and if you happy, press "Finalize".

Thanks, Alan, this looks good to me. I have pressed "Finalize".

The CSR makes promises about the Release Note. Do you think the current Release Note delivers on this promise?

https://bugs.openjdk.org/browse/JDK-8305603

eirbjo commented:
It seems the stars are getting aligned for this PR to be integrated.

None of the current approvals apply to the latest commit d04df25ea2dd505f2224bff446865a0974a8f37a, it would be good to have that `jar --help` update approved as well.

Are there any other concerns left before we can integrate?

LanceAndersen commented:
> It seems the stars are getting aligned for this PR to be integrated.
> 
> None of the current approvals apply to the latest commit [d04df25](https://github.com/openjdk/jdk/commit/d04df25ea2dd505f2224bff446865a0974a8f37a), it would be good to have that `jar --help` update approved as well.
> 
> Are there any other concerns left before we can integrate?

I think we are good to go.

eirbjo commented:
/integrate

eirbjo commented:
> I think we are good to go.

Thanks, the PR is now ready to be sponsored by a Committer.

LanceAndersen commented:
/sponsor
